components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1213 @@
+        while iters < max_iters
+            && 270.0 >= Astronomical::solar_longitude(Self::midnight_in_china(day + 1.0))
+        {
+            iters += 1;
+            day += 1.0;
+        }
+        debug_assert!(
+            iters < max_iters,
+            "Number of iterations was higher than expected"
+        );
+        day.as_rata_die()
+    }
+
+    /// Get the ISO date of the nearest Chinese New Year on or before a given ISO date
+    /// ```rust
+    /// use icu::calendar::Date;
+    /// use icu::calendar::chinese::Chinese;
+    ///
+    /// let date = Date::try_new_iso_date(2023, 6, 22).expect("Failed to initialize ISO Date");
+    /// let chinese_new_year = Chinese::chinese_new_year_on_or_before_iso(date);
+    /// assert_eq!(chinese_new_year.year().number, 2023);
+    /// assert_eq!(chinese_new_year.month().ordinal, 1);
+    /// assert_eq!(chinese_new_year.day_of_month().0, 22);
+    /// ```
+    pub fn chinese_new_year_on_or_before_iso(iso: Date<Iso>) -> Date<Iso> {
+        let iso_inner = iso.inner;
+        let fixed = Iso::fixed_from_iso(iso_inner);
+        let result_fixed = Self::chinese_new_year_on_or_before_fixed_date(fixed);
+        Iso::iso_from_fixed(result_fixed)
+    }
+
+    /// Get the fixed date of the nearest Chinese New Year on or before a given fixed date.
+    ///
+    /// Based on functions from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5396-L5405
+    pub(crate) fn chinese_new_year_on_or_before_fixed_date(date: RataDie) -> RataDie {
+        let new_year = Self::chinese_new_year_in_sui(date);
+        if date >= new_year {
+            new_year
+        } else {
+            Self::chinese_new_year_in_sui(date - 180)
+        }
+    }
+
+    /// Get a Date<Chinese> from a fixed date
+    ///
+    /// Months are calculated by iterating through the dates of new moons until finding the last month which
+    /// does not exceed the given fixed date. The day of month is calculated by subtracting the fixed date
+    /// from the fixed date of the beginning of the month.
+    ///
+    /// The calculation for `elapsed_years` in this function is based on code from _Calendrical Calculations_ by Reingold & Dershowitz.
+    /// Lisp reference code: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L5414-L5459
+    #[allow(clippy::unwrap_used)]
+    pub(crate) fn chinese_date_from_fixed(date: RataDie) -> Date<Chinese> {
+        let new_year = Self::chinese_new_year_on_or_before_fixed_date(date);
+        let elapsed_years = libm::floor(
+            1.5 - 1.0 / 12.0 + ((new_year - CHINESE_EPOCH) as f64) / MEAN_TROPICAL_YEAR,
+        );
+        let elapsed_years_int = i64_to_i32(elapsed_years as i64);
+        debug_assert!(
+            matches!(elapsed_years_int, I32Result::WithinRange(_)),
+            "Chinese year should be in range of i32"
+        );
```

Comment:
Issue: you should return "max date" or "min date" instead of debug-asserting here; see what we did in other calendars. (Can be a follow-up)